python/lsst/ap/association/packageAlerts.py

```diff
@@ -0,0 +1,247 @@
+    )
+    alertWriteLocation = pexConfig.Field(
+        dtype=str,
+        doc="Location to write alerts to.",
+        default=os.path.join(os.getcwd(), "alerts"),
+    )
+
+
+class PackageAlertsTask(pipeBase.Task):
+    """Tasks for packaging Dia and Pipelines data into Avro alert packages.
+    """
+    ConfigClass = PackageAlertsConfig
+    _DefaultName = "packageAlerts"
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.alertSchema = alertPack.Schema.from_file(self.config.schemaFile)
+        self.cutoutBBox = geom.Extent2I(self.config.cutoutSize,
+                                        self.config.cutoutSize)
+        os.makedirs(self.config.alertWriteLocation, exist_ok=True)
+
+    def run(self,
+            diaSourceCat,
+            diaObjectCat,
+            diaSrcHistory,
+            diffIm,
+            template,
+            ccdExposureIdBits):
+        """Package DiaSources/Object and exposure data into Avro alerts.
+
+        Writes Avro alerts to a location determined by the
+        ``alertWriteLocation`` configurable.
+
+        Parameters
+        ----------
+        diaSourceCat : `pandas.DataFrame`
+            New DiaSources to package. DataFrame should be indexed on
+            ``["diaObjectid", "filterName", "diaSourceId"]``
+        diaObjectCat : `pandas.DataFrame`
+            New and updated DiaObjects matched to the new DiaSources. DataFrame
+            is indexed on ``["diaObjectId"]``
+        diaSrcHistory : `pandas.DataFrame`
+            12 month history of DiaSources matched to the DiaObjects. Excludes
+            the newest DiaSource and is indexed on
+            ``["diaObjectid", "filterName", "diaSourceId"]``
+        diffIm : `lsst.afw.image.ExposureF`
+            Difference image ``diaSourceCat`` were detect in.
+        template : `lsst.afw.image.ExposureF` or `None`
+            Template image used to create the ``diffIm``.
+        ccdExposureIdBits : `int`
+            Number of bits used in the ccdVisitId.
+        """
+        alerts = []
+        self._patchDiaSources(diaSourceCat)
+        self._patchDiaSources(diaSrcHistory)
+        self._patchDiaObjects(diaObjectCat)
+        ccdVisitId = diffIm.getInfo().getVisitInfo().getExposureId()
+        for srcIndex, diaSource in diaSourceCat.iterrows():
+            # Get all diaSources for the associated diaObject.
+            diaObject = diaObjectCat.loc[srcIndex[0]]
+            if diaObject["nDiaSources"] > 1:
+                objSourceHistory = diaSrcHistory.loc[srcIndex[0]]
+            else:
+                objSourceHistory = None
+            sphPoint = geom.SpherePoint(diaSource["ra"],
+                                        diaSource["decl"],
+                                        geom.degrees)
+            diffImCutout = diffIm.getCutout(sphPoint, self.cutoutBBox)
+            templateCutout = None
+            alertId = diaSource["diaSourceId"]
```

Comment:
We are empowered to make this decision. Calling in external help if we need it.

Couple of things occur to me:

- Could we ever issue more than one alert for the same DIASource? I _think_ the answer _might_ be yes, which certainly means there can't be a one-to-one mapping between `alertId` and `diaSourceId`.
- The current schema does indeed say it's a 64 bit int, which isn't long enough to capture a UUID.
- I don't think the type of `alertId` is actually defined in the DPDD, so in theory we could make it whatever we like, including a sequence of bytes or a 128 bit int.
- If we did feel constrained to use 64 bits, but we _really_ wanted to use random numbers... we should work out what the collision probability is.
- Are we concerned about https://dba.stackexchange.com/questions/115766/should-i-use-uuid-as-well-as-id/119129#119129 ? Since the `alertId` is specifically designed to enable database queries, it seems like a valid point (but one I am too tired to think about for now).

Reply:
> If we did feel constrained to use 64 bits, but we really wanted to use random numbers... we should work out what the collision probability is.
This is the birthday problem with 2^64 birthdays. Collision probability with 1B alerts is 3%; 1B alerts is ~100 days of data. That collision probability seems waaaay too high to me.

> Are we concerned about https://dba.stackexchange.com/questions/115766/should-i-use-uuid-as-well-as-id/119129#119129 ? Since the alertId is specifically designed to enable database queries, it seems like a valid point (but one I am too tired to think about for now).

Yeah, this is a legit concern. Note that the one you linked is specific to MySQL; Postgres (for example) [doesn't hit the same page cache problems](https://stackoverflow.com/questions/34230208/uuid-primary-key-in-postgres-what-insert-performance-impact) on inserts because it just doesn't try to get as clever with row clustering on inserts. Databases built on log-structured merge trees (so, like 90% of the non-SQL stuff out there) _might_ care, but it's probably really hard to predict, and depends a lot on internals.

But I think the only alternative that would resolve this would be sequential IDs. Is there a plausible option there? It doesn't seem like we want to coordinate to come up with a sequential ordering for all alerts across CCD processors. _Maybe_ there is something clever to do where you use a 128-bit value; the first 64 bits are random and the last 64 bits are the `diaSourceId`? That would probably work better for the specific MySQL case, but who knows what effect it would have on other databases. Tough to predict.

Is insert performance really the thing to focus on, by the way? I think random read performance will be way more important for this field, but I don't know.

Reply:
{quote}
That collision probability seems waaaay too high to me.
{quote}
👍 
(But I also think that bumping to 128 bit is no big deal.)

{quote}
But I think the only alternative that would resolve this would be sequential IDs. Is there a plausible option there?
{quote}
Well, you can construct something that feels... not unreasonable to me from the information we have to hand (CCD, visit, time when the processing is taking place, host on which the processing is taking place, ...). I note your previous concerns about making these IDs opaque though.

{quote}
Is insert performance really the thing to focus on, by the way? I think random read performance will be way more important for this field, but I don't know.
{quote}
Insert performance is clearly bounded (we know how many alerts we'll be producing), while read performance isn't necessarily, so I agree with you there.

That said, it's not really clear what the load on the Alert DB will be in general. My instinct is that it's a fairly niche service, but @ebellm has thought about it much more than I have.

Reply:
> Well, you can construct something that feels... not unreasonable to me from the information we have to hand (CCD, visit, time when the processing is taking place, host on which the processing is taking place, ...).

Can you be more specific about this? How will they be densely packed and sequential?

Reply:
Not off the top of my head. :-)

But I'm assuming we're not aiming for some sort of global-sequentiality. It seems like batching by CCD/visit and ordering within that should be better than pure-random, and I'd imagine that we can find some way of packing them nicely. But I haven't though that through in any detail.

Reply:
Right, I agree that globally sequential is off the table. 

To me, the question sounds like it is whether kinda-sorta-ordered IDs are better for database performance. That's unanswerable without knowing the database and the ID scheme specifics. It's also unanswerable without knowing really what we mean by "better" for database performance, and how good is good enough. Figuring all that out sounds like an awful lot of work; my advice would be to not really worry about it. 128 bits from /dev/urandom is at least dead-simple.

Reply:
Maybe this discussion warrants it's own ticket? What I put in there was just a place holder for now and can put a TODO in and reference the ticket to decide on how to create these IDs.

Reply:
Hey @spenczar did you make a ticket for this yet? If not I can create one now.

Reply:
@morriscb here you go: https://jira.lsstcorp.org/browse/DM-24858

Reply:
Added TODO